### DISCUSSION TITLE
Embed fonts in app?

### DISCUSSION THREAD

**Discussion by wyhinton** [fltk-rs/fltk-rs#808] [Q&A] [answered]
Is it possible to embed a font as bytes in your binary, then load the font into the app?

I know we can create an ```SvgImage``` from included bytes with ```SvgImage::from_data()```, but I don't see a ```from_data()``` for ```Font```.

```rust
use fltk::{
    app::*, button::*, enums::*, prelude::*, window::*,
};

fn main(){
    let font_bytes = include_bytes!("Octicons.ttf");
    //can't load a font from bytes
    Font::load_font(font_bytes);
    let app = App::default();
    let mut win = Window::new(200, 200, 500, 300, "Event Trickle");
    let mut button= Button::new(0,0,250,200,"my button ");
    button.set_color(Color::Red);
    win.end();
    win.show();
    app.run().unwrap();
}
```

**Comment by wyhinton** [comment, marked as answer]
Looks like FLTK does not support this, see [this thread from fltk.general](https://groups.google.com/g/fltkgeneral/c/uAdg8wOLiMk).

**Reply by MoAlyousef** [reply]
Unfortunately FLTK doesn't expose that functionality. It uses the system's fonts by default. So basically loading fonts in the Rust wrapper consists of adding a font to the system and asking FLTK to read it. stb_truetype is used just to get the font name for FLTK to find it. 
It's possible to use an external library to load fonts from data but you'll have to rasterize it as an image and draw it on the widget.

**Reply by wyhinton** [reply]
Yes, I was hoping to use an icon font (like [fontawesome](https://fontawesome.com/)) in my fltk app so I could avoid having to embed an SVG for each individual icon I wanted to use. Being able to load fonts from data would be great, I much prefer embedding into the binary over bundling into an OS-specific app package. The later seems far more prone to error (and requires a fancy installer).

> It's possible to use an external library to load fonts from data 

I didn't realize that FLTK had no means of loading a font from a path. You're saying you added this to fltk-rs as a utility? If you've figured out how to inject font's loaded from paths into the system font's a run time, seems like we'd just need to find a way to convert the embedded font bytes into a font, then inject this font into the system fonts.

>but you'll have to rasterize it as an image and draw it on the widget.

Quite the workaround. So basically each glyph becomes a raster image? There's no way to "reassemble" these raster images back into something you could use like a font (font sizes, angled text, colored fonts.)? Do you think it'd be hard to implement fonts from data in FLTK? Aren't several font file types already in a binary format (at least .ttf, I believe)?

**Reply by MoAlyousef** [reply]
> seems like we'd just need to find a way to convert the embedded font bytes into a font, then inject this font into the system fonts

This can be done by taking the include_bytes! data, writing it to a temporary file, loading the font then deleting the file. Unfortunately trying to plug into FLTK's font drawing otherwise (without file i/o) is not possible AFAIK.
```rust
use fltk::{
    app::*, button::*, enums::*, prelude::*, window::*,
};

struct MyFont {}

impl MyFont {
    pub fn from_bytes(font: &[u8]) {
        let mut temp_file = std::env::temp_dir();
        temp_file.push("myfile.ttf");
        std::fs::write(&temp_file, font).unwrap();
        let name = Font::load_font(temp_file).unwrap();
        Font::set_font(Font::Helvetica, &name); // swap with the default font
    }
}

fn main(){
    let font_bytes = include_bytes!("DejaVuSansMono-Bold.ttf");
    MyFont::from_bytes(font_bytes);
    let app = App::default();
    let mut win = Window::new(200, 200, 500, 300, "Event Trickle");
    let mut button= Button::new(0,0,250,200,"my button ");
    button.set_color(Color::Red);
    win.end();
    win.show();
    app.run().unwrap();
}
```

>  There's no way to "reassemble" these raster images back into something you could use like a font (font sizes, angled text, colored fonts.)?

There are several drawing libraries which can do advanced font drawing like plotters, but I think it uses font-kit, which like FLTK expects the font to be in the system. 
Here's an example using imageproc which can draw text onto a transparent image buffer and which uses rusttype which allows loading fonts without them being in the system:
```rust
use ::image::{Rgba, RgbaImage};
use fltk::{prelude::*, *};
use imageproc::drawing::{draw_text_mut, text_size};
use rusttype::{Font, Scale};

const FONT: &[u8] = include_bytes!("DejaVuSans.ttf");

pub struct MyFrame {
    #[allow(dead_code)]
    frame: button::Button,
}

impl MyFrame {
    pub fn new(x: i32, y: i32, w: i32, h: i32, title: &str) -> Self {
        let mut frame = button::Button::new(x, y, w, h, None);
        let mut image = RgbaImage::new(w as _, h as _); // Transparent by default

        let font = Font::try_from_bytes(FONT).unwrap();

        let height = h as f32 / 4.;
        let scale = Scale {
            x: height,
            y: height,
        };
        let size = text_size(scale, &font, title);
        draw_text_mut(
            &mut image,
            Rgba([0u8, 0u8, 0u8, 255u8]),
            (w - size.0) / 2,
            (h - size.1) / 2,
            scale,
            &font,
            title,
        );
        frame.draw(move |f| {
            let mut rgb = image::RgbImage::new(&image, w, h, enums::ColorDepth::Rgba8).unwrap();
            rgb.draw(f.x(), f.y(), f.w(), f.h());
        });
        Self { frame }
    }
}

fn main() {
    let a = app::App::default();
    let mut win = window::Window::default().with_size(400, 300);
    MyFrame::new(10, 10, 100, 100, "Hello");
    win.end();
    win.show();
    a.run().unwrap();
}
```
```toml
[dependencies]
fltk = "1"
imageproc = { git = "https://github.com/image-rs/imageproc" } # we use master for the text_size method.
image = {version = "0.23.14", default-features=false }
rusttype = "0.9"
```
